onnxruntime/core/providers/dnnl/dnnl_custom_threadpool.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#include "dnnl_threadpool_iface.hpp"
+#include "core/platform/threadpool.h"
+#include "core/platform/EigenNonBlockingThreadPool.h"
+
+// oneDNN limitations using eigen
+#if !(EIGEN_WORLD_VERSION == 3 && EIGEN_MAJOR_VERSION == 3) && defined(DNNL_EIGEN_THREAD)
+#error Unsupported Eigen version (need 3.3.x)
+#endif
+
+// To improve readability
+using OneDnnlThreadPoolIface = dnnl::threadpool_interop::threadpool_iface;
+using ORTThreadPool = onnxruntime::concurrency::ThreadPool;
+using EigenThreadPool = Eigen::ThreadPool;
+
+
+inline int calc_num_threads() {
+  int num_treads = 0;
+#if _WIN32
+  // Indeed 64 should be enough. However, it's harmless to have a little more.
+  SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
+  DWORD returnLength = sizeof(buffer);
+  if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
+    num_treads = std::thread::hardware_concurrency();
+  } else {
+    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
+    for (int i = 0; i != count; ++i) {
+      if (buffer[i].Relationship == RelationProcessorCore) {
+        ++num_treads;
+      }
+    }
+  }
+#endif
+  if (!num_treads)
+    num_treads = std::thread::hardware_concurrency();
+
+  return num_treads;
+}
+
+template <typename T>
+struct remove_reference {
+    typedef T type;
+};
+template <typename T, typename U>
+inline void balance211(T n, U team, U tid, T& n_start, T& n_end) {
+  T n_min = 1;
+  T& n_my = n_end;
+  if (team <= 1 || n == 0) {
+    n_start = 0;
+    n_my = n;
+  } else if (n_min == 1) {
+    // team = T1 + T2
+    // n = T1*n1 + T2*n2  (n1 - n2 = 1)
+    T c_team = (T)team;
+    assert(c_team);
+    T n1 = static_cast<typename remove_reference<T>::type>((n + c_team - 1) / c_team);
+    T n2 = n1 - 1;
+    T T1 = n - n2 * c_team;
+    n_my = (T)tid < T1 ? n1 : n2;
+    n_start = (T)tid <= T1 ? tid * n1 : T1 * n1 + ((T)tid - T1) * n2;
+  }
+
+  n_end += n_start;
+}
+
+// Define base beheaviour for all custom threadpools
+template <class T>
+class OneDnnlThreadPool : public OneDnnlThreadPoolIface {
+
+ public:
+  explicit OneDnnlThreadPool(T* threadpool) : inner_treadpool_(threadpool) {}
+  uint64_t get_flags() const override { return ASYNCHRONOUS; }
+
+ protected:
+  T* inner_treadpool_;
+};
+
+
+class OneDnnlORTThreadPool : public OneDnnlThreadPool<ORTThreadPool> {
+ public:
+  OneDnnlORTThreadPool(ORTThreadPool* threadpool) : OneDnnlThreadPool<ORTThreadPool>(threadpool) {}
+  int get_num_threads() const override { return inner_treadpool_->GetNumThreads(); }
+  bool get_in_parallel() const override { return inner_treadpool_->GetCurrentThreadId() != -1; }
+  uint64_t get_flags() const override { return !ASYNCHRONOUS; }
```

Comment:
No, this function is part of the oneDNN threadpool interface, what it does is it signals if the parallel for is going to be asynchronous (which means oneDNN will call _parallel_for()_ and when the function finishes threads can be executing the function so oneDNN needs to implement its own waiting)  or synchronous (which means that when  _parallel_for()_ returns the function is finished, so the waiting is handled by the custom threadpool, which is our case)

Reply:
> No, this function is part of the oneDNN threadpool interface, what it does is it signals if the parallel for is going to be asynchronous (which means oneDNN will call _parallel_for()_ and when the function finishes threads can be executing the function so oneDNN needs to implement its own waiting) or synchronous (which means that when _parallel_for()_ returns the function is finished, so the waiting is handled by the custom threadpool, which is our case)
> 
> As part of the default interface they defined the _ASYNCHRONOUS_=1 attribute as the default, so I just thought implementing the _get_flags()_ by using that member would be good for readability, but I can create a _SYNCHRONOUS_ member if you prefer.

I thought it might be something like that. Thanks for verifying what was the case.